flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1882,9 +1882,11 @@ int getNumQueuedRequests() {
     public void reportStats(long id, ExecutionAttemptID attemptId, CheckpointMetrics metrics)
             throws CheckpointException {
         if (statsTracker != null) {
-            attemptMappingProvider
-                    .getVertex(attemptId)
-                    .ifPresent(ev -> statsTracker.reportIncompleteStats(id, ev, metrics));
+            synchronized (lock) {
```

Comment:
I am a bit concerned we do synchronize too much. I don't like reporting the stats happen inside of the lock as well. We should not delay the checkpointing process for the sake of stats. The stats tracker already supports dirty reads, which we should not discard I believe. 

What do you think about fixing it inside of the `ExecutionAttemptMappingProvider`? I think we could improve that class to be thread safe.

Reply:
I also had such idea but I decided to avoid the second lock. But maybe you are right and it doesn't make sense to report stats under the lock.